### PR TITLE
Changes User.last_attempt to return a UserLevel rather than an Activity.

### PR DESCRIPTION
**Pull request by ashercodeorg**
Inspecting the views that ultimately use the return of `last_attempt`, all of them require the method `level_source` to be defined on the return. Since this is still the case, none of the views seem to need updating.

After this merged is into staging (test), I plan to bash on it some to confirm nothing is broken. Bashing assistance is certainly welcome.

### CODE REVIEW

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -518,14 +518,14 @@ def unpassed_progression_level?(script_level, user_levels)
...
-  # Returns the most recent (via created_at) activity for the specified level.
-  # TODO(asher): Change this to use UserLevel rather than Activity.
+  # Returns the most recent (via updated_at) user_level for the specified
+  # level.
   def last_attempt(level)
-    Activity.where(user_id: self.id, level_id: level.id).order('id desc').first
+    UserLevel.where(user_id: self.id, level_id: level.id).
+      order('updated_at DESC').
+      first
   end
```
Comment (ashercodeorg) [inline dashboard/app/models/user.rb:527]: Note that we might want to refine this implementation, recalling that `Activity` has no context of a `Script`, whereas `UserLevel` does have such context. We could therefore pass in an optional script, filtering by it (if present).

**Comment by ashercodeorg** [comment]
TODO (followup PRs):
- [ ] Rename `last_attempt` method to `last_user_level` or otherwise.
- [ ] Rename `last_attempt` variables scattered around (controllers and views) to `last_user_level` or otherwise.
- [ ] Determine whether we want to add a `script` context to `last_attempt`, adding it if so.

**Comment by ashercodeorg** [comment]
CircleCI failure is legitimate, the second commit fixes it.

**Comment by ashercodeorg** [comment]
(rebased and force pushed to merge with staging)

File: dashboard/app/models/user.rb [hunk 2]
```diff
@@ -518,14 +518,16 @@ def unpassed_progression_level?(script_level, user_levels)
...
 
-  # Returns the most recent (via created_at) activity for the specified level.
-  # TODO(asher): Change this to use UserLevel rather than Activity.
+  # Returns the most recent (via updated_at) user_level for the specified
+  # level.
   def last_attempt(level)
-    Activity.where(user_id: self.id, level_id: level.id).order('id desc').first
+    UserLevel.where(user_id: self.id, level_id: level.id).
+      order('updated_at DESC').
```
Comment (Hamms) [inline dashboard/app/models/user.rb:525]: is this indexed?

File: dashboard/test/controllers/api_controller_test.rb [hunk 3]
```diff
@@ -30,6 +30,9 @@ class ApiControllerTest < ActionController::TestCase
     @student_flappy_1 = create(:follower, section: @flappy_section).student_user
     @student_flappy_1.backfill_user_scripts
     @student_flappy_1.reload
+
+    # Seed the RNG deterministically.
+    srand 1
```
Comment (Hamms) [inline dashboard/test/controllers/api_controller_test.rb]: why is this necessary?

Comment (ashercodeorg) [reply dashboard/test/controllers/api_controller_test.rb]: For the anonymous assessment question. Note that I moved it from that specific test to the setup. Happy to put it back.

File: dashboard/app/models/user.rb (hunk 2 above)
Comment (ashercodeorg) [reply dashboard/app/models/user.rb:525]: Good call (nice catch!). Let me rethink this.

**Comment by Hamms** [comment]
LGTM, modulo a couple questions

File: dashboard/app/models/user.rb (hunk 2 above)
Comment (ashercodeorg) [reply dashboard/app/models/user.rb:525]: Unfortunately this wasn't caught in my change to `last_attempt_for_any`, e.g., in #10120.

File: dashboard/test/controllers/api_controller_test.rb (hunk 3 above)
Comment (Hamms) [reply dashboard/test/controllers/api_controller_test.rb]: I'd vote to either put it back or preserve the original comment

File: dashboard/app/models/user.rb (hunk 2 above)
Comment (ashercodeorg) [reply dashboard/app/models/user.rb:525]: Oh, never mind. The lookup by `user_id` and `level_id` is indexed. As we expect very few of them, the `order` is fine without index.

Comment (Hamms) [reply dashboard/app/models/user.rb:525]: Oh, of course! Delightful!

File: dashboard/test/controllers/api_controller_test.rb (hunk 3 above)
Comment (ashercodeorg) [reply dashboard/test/controllers/api_controller_test.rb]: Done (mostly both).
